src/main/java/org/apache/cassandra/sidecar/routes/HealthService.java

```diff
@@ -18,55 +18,83 @@
 
 package org.apache.cassandra.sidecar.routes;
 
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
 import com.google.common.collect.ImmutableMap;
 
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.Host;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
 import io.netty.handler.codec.http.HttpHeaderValues;
 import io.netty.handler.codec.http.HttpResponseStatus;
 import io.vertx.core.http.HttpHeaders;
 import io.vertx.core.json.Json;
 import io.vertx.core.logging.Logger;
 import io.vertx.core.logging.LoggerFactory;
 import io.vertx.ext.web.RoutingContext;
+import org.apache.cassandra.sidecar.CQLSession;
+import org.apache.cassandra.sidecar.Configuration;
 
-import java.util.concurrent.Executors;
-import java.util.concurrent.ScheduledExecutorService;
-import java.util.concurrent.TimeUnit;
-import java.util.function.Supplier;
-
-public class HealthService
+/**
+ * Tracks health check[s] and provides a REST response that should match that defined by api.yaml
+ */
+@Singleton
+public class HealthService implements Host.StateListener
 {
     private static final Logger logger = LoggerFactory.getLogger(HealthService.class);
-    private final int CHECK_PERIOD_MS;
+    private final int checkPeriodMs;
     private final Supplier<Boolean> check;
+    private final CQLSession session;
+    private volatile boolean registered = false;
 
     private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
     private volatile boolean lastKnownStatus = false;
 
-    public HealthService(int checkPeriodMillis, Supplier<Boolean> check)
+    @Inject
+    public HealthService(Configuration config, HealthCheck check, CQLSession session)
     {
-        this.CHECK_PERIOD_MS = checkPeriodMillis;
+        this.checkPeriodMs = config.getHealthCheckFrequencyMillis();
+        this.session = session;
         this.check = check;
     }
 
-    synchronized public void start()
+    public synchronized void start()
     {
         logger.info("Starting health check");
-        executor.scheduleWithFixedDelay(this::refreshNow, 0, CHECK_PERIOD_MS, TimeUnit.MILLISECONDS);
+        maybeRegisterHostListener();
+        executor.scheduleWithFixedDelay(this::refreshNow, 0, checkPeriodMs, TimeUnit.MILLISECONDS);
     }
 
-    synchronized public void refreshNow()
+    public synchronized void refreshNow()
     {
         try
         {
             lastKnownStatus = this.check.get();
+            maybeRegisterHostListener();
         }
         catch (Exception e)
         {
             logger.error("Error while performing health check", e);
         }
     }
 
-    synchronized public void stop()
+    private synchronized void maybeRegisterHostListener()
+    {
+        if (!registered)
+        {
+            if (session != null && session.getLocalCql().isPresent())
```

Comment:
Using optional instead of a null check doesn't add any value. See: 
https://stackoverflow.com/a/40801034